pycheribuild/projects/run_qemu.py

```diff
@@ -413,28 +418,71 @@ def setup_config_options(cls, **kwargs):
                  "Useful when building and running on separate machines.")
+        cls.list_kernels = cls.add_bool_option("list-kernels", show_help=False,
+                                               help="List available kernel configs to run and exit")
+        cls.kernel_config = cls.add_config_option(
+            "kernel-config", show_help=True,
+            help="Select the kernel to run by specifying the kernel build configuration name."
+                 "The list of available kernel configurations is given by --list-kernels")
+        cls.kernel_abi = cls.add_config_option(
+            "kernel-abi", show_help=True,
+            kind=KernelABI, enum_choices=[KernelABI.HYBRID, KernelABI.PURECAP],
+            help="Select extra kernel variant with the given ABI to run.")
 
     def __init__(self, config: CheriConfig, source_class: "typing.Type[BuildFreeBSD]" = None,
                  disk_image_class: "typing.Type[BuildFreeBSDImage]" = None, needs_disk_image=True):
         super().__init__(config)
         if source_class is None and disk_image_class is not None:
             # noinspection PyProtectedMember
-            source_class = disk_image_class.get_instance(self).source_project
-        self.source_class = source_class
-        self.current_kernel = source_class.get_installed_kernel_path(self)
+            self.source_project = disk_image_class.get_instance(self).source_project
+        else:
+            self.source_project = source_class.get_instance(self)
+
+        if self.kernel_config:
+            if self.kernel_config not in self._valid_kernel_configs():
+                self.fatal("Selected kernel configuration", self.kernel_config, "is not available")
+                self._list_kernel_configs()
+        else:
+            config_filters = {}
+            if self.kernel_abi:
+                if self.crosscompile_target.is_hybrid_or_purecap_cheri():
+                    config_filters["kABI"] = self.kernel_abi
+                else:
+                    self.warning("Can not select kernel ABI to run for non-CHERI target, ignoring --kernel-abi")
+            self.kernel_config = self.source_project.default_kernel_config(ConfigPlatform.QEMU, **config_filters)
+
+        self.current_kernel = self.source_project.get_kernel_install_path(self.kernel_config)
+
+        if self.qemu_options.can_boot_kernel_directly:
+            self._project_specific_options += ["-append", "kern.module_path={}".format(self.current_kernel.parent)]
         if hasattr(source_class, "get_rootfs_dir"):
             # noinspection PyCallingNonCallable
             self.rootfs_path = source_class.get_rootfs_dir(self, config=config)
         if needs_disk_image:
             self.disk_image = disk_image_class.get_instance(self).disk_image_path
 
+    def _valid_kernel_configs(self):
+        return self.source_project.get_kernel_configs(platform=ConfigPlatform.QEMU)
```

Comment:
Should we just list all files `sys/<arch>/conf/` that don't contain a "." to get the full list?

Reply:
There are configs in `sys/<arch>/conf` that we do not use or support in cheribuild and we never build. This would be sensible if we generated the set of kernel configurations by reading files in `sys/<arch>/conf` in the first place, which is possible, but I just went the other way around in the generator to avoid having a blacklist of files. I can try to have the generator read config files in `sys/<arch>` but would leave this call as is, so that it does not care about how we get configuration names

Reply:
Sorry, I should have been a bit clearer in my comment: I meant printing the files in `sys/<arch>/conf` only for the --list-kernels option (possibly indicating which ones are automatically used by cheribuild), so that you can see the full list of strings that you can pass as the --kernel-config override. Does that make more sense?

Reply:
Ah yes that can be done, however I think that in the cheribsd target it should list the kernels that can be given as default kernel overriede in --cheribsd/kernel-config, instead in qemu it should list the kernels that have been built and can be run (yet with the --qemu/kernel-config override, possibly changing the name of the qemu option to avoid confusion)